### PR TITLE
Create an Expr class to parse and save the AST, allowing retrieving the list of names

### PR DESCRIPTION
**Pull request by charlax**
This PR is based on #35. It adds some extra functionality:

* You can now save the generated AST by using the `Expr` class.
* You can now get the list of first-level names defined in the AST. This is useful if you want to get an idea of the names you'll have to provide or to define the dependencies of an expression.

The current implementation is not super efficient because an evaluator instance is created for each evaluation. Let me know if this should be fixed in this PR.

### CODE REVIEW

**Comment by danthedeckie** [comment]
Hi @charlax ,

Thanks!  I merged in #35, which means this PR is now conflicting due to those changes being made here too, so I can't merge this in as is.

What exactly is the use-case for being able to inspect the AST / get the used names?
I can see the point of parsing once, then evaluating multiple times with different names={} - as in https://github.com/danthedeckie/simpleeval/issues/25#issuecomment-277796587 .

When I have time to write all the tests, I'd really like to split out the parse() and eval() methods as @dave-nm mentioned.  The AST would then be inspectable, with the visitor as you do here.

I just don't quite see a general use-case for this though?

**Comment by charlax** [comment]
For sure. My use case is the following (I'm simplifying a bit).

I have a form with multiple steps. Each step gathers different fields and has a formula attached. The formula might use fields from previous steps. I wrote a script that validates the form's consistency, i.e. that there is no step formula that depends on fields that have not yet been filled by the user.

In simpler terms, knowing what names are used in a formula allows me to check if those names are available.

Let me know if you're interested in this PR, and I'll fix the conflicts.
